File: app/src/main/java/es/etologic/mahjongscoring2/app/main/combinations/CombinationsActivity.java
package es.etologic.mahjongscoring2.app.main.combinations;

import android.app.SearchManager;
import android.arch.lifecycle.ViewModelProviders;
import android.content.Context;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.SearchView;
import android.support.v7.widget.Toolbar;
import android.view.Menu;
import android.view.MenuItem;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.Unbinder;
import es.etologic.mahjongscoring2.R;
import es.etologic.mahjongscoring2.app.base.BaseActivity;

import static es.etologic.mahjongscoring2.app.model.ShowState.SHOW;

public class CombinationsActivity extends BaseActivity {

    //Fields
    @BindView(R.id.toolbarCombinations) Toolbar toolbar;
    @BindView(R.id.recyclerViewCombinations) RecyclerView recyclerView;
    private Unbinder unbinder;
    @Inject CombinationsViewModelFactory combinationsViewModelFactory;
    private CombinationsViewModel viewModel;
    private CombinationsRvAdapter rvAdapter;

    //Lifecycle

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        setTheme(R.style.AppTheme);
        super.onCreate(savedInstanceState);
        setContentView(R.layout.combinations_activity);
        unbinder = ButterKnife.bind(this);
        setupToolbar();
        setupRecyclerView();
        setupViewModel();
        viewModel.getAll();
    }

    private void setupToolbar() {
        setSupportActionBar(toolbar);
        if (getSupportActionBar() != null) {
            getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        }
    }
    public void setupRecyclerView() {
        recyclerView.setHasFixedSize(true);
        final LinearLayoutManager layoutManager = new LinearLayoutManager(this);
        recyclerView.setLayoutManager(layoutManager);
        rvAdapter = new CombinationsRvAdapter();
        recyclerView.setAdapter(rvAdapter);
    }

    private void setupViewModel() {
        viewModel = ViewModelProviders.of(this, combinationsViewModelFactory).get(CombinationsViewModel.class);
        viewModel.getFilteredCombinations().observe(this, rvAdapter::setCombinations);
    }
    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.combinations_menu, menu);
        SearchManager searchManager = (SearchManager) getSystemService(Context.SEARCH_SERVICE);
        SearchView searchView = (SearchView) menu.findItem(R.id.action_search_combination).getActionView();
        if (searchManager != null) {
            searchView.setSearchableInfo(searchManager.getSearchableInfo(getComponentName()));
            searchView.setOnQueryTextListener(new SearchView.OnQueryTextListener() {
                @Override
                public boolean onQueryTextSubmit(String query) {
                    return false;
                }

                @Override
                public boolean onQueryTextChange(String newText) {
                    viewModel.searchCombination(newText);
                    return true;
                }
            });
        }
        return true;
    }
    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                onBackPressed();
                return true;
            case R.id.action_toggle_combination_explanation:
                MenuItem menuItem = toolbar.getMenu().getItem(1);
                menuItem.setIcon(rvAdapter.toggleImageOrDescription() == SHOW ?
                        R.drawable.ic_library_books_white_24dp :
                        R.drawable.ic_photo_library_white_24dp);
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }
    @Override
    public void onDestroy() {
        unbinder.unbind();
        super.onDestroy();
    }
}

File: app/src/main/java/com/etologic/mahjongscoring2/app/main/combinations/CombinationsActivity.kt
/*
    Copyright (C) 2020  Ernesto Vega de la Iglesia Soria
 
    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.
    
    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.
    
    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <https://www.gnu.org/licenses/>.
*/
package com.etologic.mahjongscoring2.app.main.combinations

import android.app.SearchManager
import android.content.Context
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.view.Menu
import android.view.MenuItem
import androidx.appcompat.widget.SearchView
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.LinearLayoutManager
import com.etologic.mahjongscoring2.R
import com.etologic.mahjongscoring2.app.base.BaseActivity
import com.etologic.mahjongscoring2.app.model.ShowState.SHOW
import com.etologic.mahjongscoring2.databinding.CombinationsActivityBinding
import javax.inject.Inject

class CombinationsActivity : BaseActivity() {

    private lateinit var binding: CombinationsActivityBinding

    @Inject
    internal lateinit var combinationsViewModelFactory: CombinationsViewModelFactory
    internal var viewModel: CombinationsViewModel? = null

    @Inject
    internal lateinit var rvAdapter: CombinationsRvAdapter

    override val onBackBehaviour = { finish() }

    override fun onCreate(savedInstanceState: Bundle?) {
        setTheme(R.style.AppTheme)
        super.onCreate(savedInstanceState)
        binding = CombinationsActivityBinding.inflate(layoutInflater)
        val view = binding.root
        setContentView(view)
    }

    override fun onPostCreate(savedInstanceState: Bundle?) {
        super.onPostCreate(savedInstanceState)
        setupToolbar()
        setupRecyclerView()
        setupViewModel()
        viewModel?.getAll()
    }

    private fun setupToolbar() {
        setSupportActionBar(binding.toolbarCombinations)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)
    }

    private fun setupRecyclerView() {
        binding.recyclerViewCombinations.setHasFixedSize(true)
        val layoutManager = LinearLayoutManager(this)
        binding.recyclerViewCombinations.layoutManager = layoutManager
        val handler = Handler(Looper.getMainLooper())
        handler.post { binding.recyclerViewCombinations.adapter = rvAdapter }
    }

    private fun setupViewModel() {
        viewModel = ViewModelProvider(this, combinationsViewModelFactory)[CombinationsViewModel::class.java]
        viewModel?.getFilteredCombinations()?.observe(this) { rvAdapter.setCombinations(it) }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.combinations_menu, menu)
        val searchManager = getSystemService(Context.SEARCH_SERVICE) as SearchManager
        val searchView = menu.findItem(R.id.action_search_combination).actionView as SearchView
        searchView.setSearchableInfo(searchManager.getSearchableInfo(componentName))
        searchView.setOnQueryTextListener(object : SearchView.OnQueryTextListener {
            override fun onQueryTextSubmit(query: String): Boolean {
                return false
            }

            override fun onQueryTextChange(newText: String): Boolean {
                viewModel?.searchCombination(newText)
                return true
            }
        })
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            android.R.id.home -> {
                finish()
                return true
            }

            R.id.action_toggle_combination_explanation -> {
                val menuItem = binding.toolbarCombinations.menu?.getItem(1)
                menuItem?.setIcon(
                    if (rvAdapter.toggleImageOrDescription() === SHOW)
                        R.drawable.ic_photo_library_white_24dp
                    else
                        R.drawable.ic_library_books_white_24dp
                )
                return true
            }

            else -> return super.onOptionsItemSelected(item)
        }
    }
}
